File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/fagsaker/FagsakController.java
package no.nav.melosys.tjenester.gui.fagsaker;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.Behandlingsresultat;
import no.nav.melosys.domain.Fagsak;
import no.nav.melosys.domain.kodeverk.Aktoersroller;
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.service.behandling.BehandlingsresultatService;
import no.nav.melosys.service.mottatteopplysninger.MottatteOpplysningerService;
import no.nav.melosys.service.persondata.PersondataFasade;
import no.nav.melosys.service.registeropplysninger.OrganisasjonOppslagService;
import no.nav.melosys.service.sak.*;
import no.nav.melosys.service.saksopplysninger.SaksopplysningerService;
import no.nav.melosys.service.tilgang.Aksesskontroll;
import no.nav.melosys.sikkerhet.context.SubjectHandler;
import no.nav.melosys.tjenester.gui.dto.*;
import no.nav.melosys.tjenester.gui.dto.periode.PeriodeDto;
import no.nav.security.token.support.core.api.Protected;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.context.WebApplicationContext;

import static no.nav.melosys.domain.util.MottatteOpplysningerUtils.hentLand;
import static no.nav.melosys.domain.util.MottatteOpplysningerUtils.hentPeriode;

@Protected
@RestController
@RequestMapping("/fagsaker")
@Api(tags = {"fagsaker"})
@Scope(value = WebApplicationContext.SCOPE_REQUEST)
public class FagsakController {
    private static final Logger log = LoggerFactory.getLogger(FagsakController.class);
    private static final String UKJENT_NAVN = "UKJENT";

    private final FagsakService fagsakService;
    private final OpprettSak opprettSak;
    private final EndreSakService endreSakService;
    private final Aksesskontroll aksesskontroll;
    private final MottatteOpplysningerService mottatteOpplysningerService;
    private final BehandlingsresultatService behandlingsresultatService;
    private final PersondataFasade persondataFasade;
    private final SaksopplysningerService saksopplysningerService;
    private final OrganisasjonOppslagService organisasjonOppslagService;
    private final OpprettBehandlingForSak opprettBehandlingForSak;
    private final FerdigbehandleService ferdigbehandleService;

    public FagsakController(FagsakService fagsakService,
                            Aksesskontroll aksesskontroll,
                            MottatteOpplysningerService mottatteOpplysningerService,
                            OpprettSak opprettSak,
                            EndreSakService endreSakService,
                            BehandlingsresultatService behandlingsresultatService,
                            PersondataFasade persondataFasade,
                            SaksopplysningerService saksopplysningerService,
                            OrganisasjonOppslagService organisasjonOppslagService,
                            OpprettBehandlingForSak opprettBehandlingForSak,
                            FerdigbehandleService ferdigbehandleService) {
        this.fagsakService = fagsakService;
        this.aksesskontroll = aksesskontroll;
        this.mottatteOpplysningerService = mottatteOpplysningerService;
        this.opprettSak = opprettSak;
        this.endreSakService = endreSakService;
        this.behandlingsresultatService = behandlingsresultatService;
        this.persondataFasade = persondataFasade;
        this.saksopplysningerService = saksopplysningerService;
        this.organisasjonOppslagService = organisasjonOppslagService;
        this.opprettBehandlingForSak = opprettBehandlingForSak;
        this.ferdigbehandleService = ferdigbehandleService;
    }

    @GetMapping("/{saksnr}")
    @ApiOperation(value = "Henter en sak med et gitt saksnummer", notes = ("Spesifikke saker kan hentes via saksnummer."))
    public ResponseEntity<FagsakDto> hentFagsak(@PathVariable("saksnr") String saksnummer) {
        Fagsak fagsak = fagsakService.hentFagsak(saksnummer);
        aksesskontroll.autoriserSakstilgang(fagsak);
        FagsakDto fagsakDto = tilFagsakDto(fagsak);
        log.info("Henting av sak {} ({})", fagsakDto.getSaksnummer(), fagsakDto.getGsakSaksnummer());
        return ResponseEntity.ok(fagsakDto);
    }

    @PostMapping
    @ApiOperation(value = "Oppretter en ny sak.")
    public ResponseEntity<Void> opprettNySak(@RequestBody OpprettSakDto opprettSakDto) {
        if (opprettSakDto.getBrukerID() == null && opprettSakDto.getVirksomhetOrgnr() == null) {
            throw new FunksjonellException("BrukerID eller organisasjonsnummer trengs for å opprette en sak.");
        }
        if (opprettSakDto.getBrukerID() != null) {
            aksesskontroll.autoriserFolkeregisterIdent(opprettSakDto.getBrukerID());
        }

        if (opprettSakDto.getOppgaveID() == null) {
            opprettSak.opprettNySakOgBehandling(opprettSakDto);
        } else {
            opprettSak.opprettNySakOgBehandlingFraOppgave(opprettSakDto);
        }

        return ResponseEntity.noContent().build();
    }

    @PostMapping("/{saksnr}/behandlinger")
    @ApiOperation(value = "Oppretter en ny behandling for sak.")
    public ResponseEntity<Void> opprettNyBehandlingForSak(@PathVariable("saksnr") String saksnummer, @RequestBody OpprettSakDto opprettSakDto) {
        if (opprettSakDto.getBrukerID() != null) {
            aksesskontroll.autoriserFolkeregisterIdent(opprettSakDto.getBrukerID());
        }

        opprettBehandlingForSak.opprettBehandling(saksnummer, opprettSakDto);

        return ResponseEntity.noContent().build();
    }

    @PutMapping("/{saksnr}")
    @ApiOperation(value = "Endre en sak.")
    public ResponseEntity<Void> endreFagsak(@PathVariable("saksnr") String saksnummer, @RequestBody EndreSakDto endreDto) {
        log.debug("Saksbehandler {} ber om å endre fagsak {} med sakstype {}, sakstema {}",
            SubjectHandler.getInstance().getUserID(), saksnummer, endreDto.getSakstype(), endreDto.getSakstema());
        aksesskontroll.autoriserSakstilgang(saksnummer);

        if (endreDto.getBehandlingstype() == Behandlingstyper.ÅRSAVREGNING && endreDto.getBehandlingID() != null) {
            endreSakService.endreÅrsavregningOppsummering(endreDto.getBehandlingID(), endreDto.getBehandlingsstatus(), endreDto.getMottaksdato());
        } else {
            endreSakService.endre(saksnummer, endreDto.getSakstype(), endreDto.getSakstema(), endreDto.getBehandlingstema(),
                endreDto.getBehandlingstype(), endreDto.getBehandlingsstatus(), endreDto.getMottaksdato());
        }

        return ResponseEntity.noContent().build();
    }

    @PostMapping("/sok")
    @ApiOperation(
        value = "Søk etter saker på ident eller saksnummer eller orgnr",
        notes = ("Saker knyttet til en bruker søkes via fødselsnummer eller d-nummer. Saker knyttet til en organisasjon søkes via organisasjonsnummer."),
        response = FagsakOppsummeringDto.class,
        responseContainer = "List")
    public List<FagsakOppsummeringDto> hentFagsaker(@RequestBody FagsakSokDto fagsakSokDto) {

        if (StringUtils.isNotEmpty(fagsakSokDto.ident())) {
            aksesskontroll.auditAutoriserFolkeregisterIdent(fagsakSokDto.ident(), "Søk på person med ident. Oversikt over saker og behandlinger.");
            return tilFagsakOppsummeringDtoer(fagsakService.hentFagsakerMedAktør(Aktoersroller.BRUKER, fagsakSokDto.ident()));
        } else if (StringUtils.isNotEmpty(fagsakSokDto.saksnummer())) {
            Optional<Fagsak> fagsak = fagsakService.finnFagsakFraSaksnummer(fagsakSokDto.saksnummer());
            if (fagsak.isPresent()) {
                aksesskontroll.auditAutoriserSakstilgang(fagsak.get(), "Søk på sak med saksnummer. Oversikt over saker og behandlinger.");
                return tilFagsakOppsummeringDtoer(Collections.singletonList(fagsak.get()));
            }
        } else if (StringUtils.isNotEmpty(fagsakSokDto.orgnr())) {
            return tilFagsakOppsummeringDtoer(fagsakService.hentFagsakerMedOrgnr(Aktoersroller.VIRKSOMHET, fagsakSokDto.orgnr()));
        }

        return Collections.emptyList();
    }

    @PutMapping("/{behandlingID}/ferdigbehandle")
    @ApiOperation("Avslutt behandling med Ferdigbehandlet som resultat og oppdatere saksstatus")
    public ResponseEntity<Void> ferdigbehandleSak(@PathVariable("behandlingID") long behandlingID) {
        log.info("Saksbehandler {} ber om å ferdigbehandle behandling {}", SubjectHandler.getInstance().getUserID(), behandlingID);
        aksesskontroll.autoriserSkriv(behandlingID);

        ferdigbehandleService.ferdigbehandle(behandlingID);

        return ResponseEntity.noContent().build();
    }


    private FagsakDto tilFagsakDto(Fagsak fagsak) {
        FagsakDto fagsakDto = new FagsakDto();
        fagsakDto.setSaksnummer(fagsak.getSaksnummer());
        fagsakDto.setGsakSaksnummer(fagsak.getGsakSaksnummer());
        fagsakDto.setSakstema(fagsak.getTema());
        fagsakDto.setSakstype(fagsak.getType());
        fagsakDto.setSaksstatus(fagsak.getStatus());
        fagsakDto.setRegistrertDato(fagsak.getRegistrertDato());
        fagsakDto.setEndretDato(fagsak.getEndretDato());
        fagsakDto.setHovedpartRolle(fagsak.getHovedpartRolle());

        return fagsakDto;
    }

    private List<FagsakOppsummeringDto> tilFagsakOppsummeringDtoer(Iterable<Fagsak> saker) {
        List<FagsakOppsummeringDto> fagsakListe = new ArrayList<>();
        for (Fagsak fagsak : saker) {
            FagsakOppsummeringDto fagsakOppsummeringDto = new FagsakOppsummeringDto();
            fagsakOppsummeringDto.setSaksnummer(fagsak.getSaksnummer());
            fagsakOppsummeringDto.setSakstema(fagsak.getTema());
            fagsakOppsummeringDto.setSakstype(fagsak.getType());
            fagsakOppsummeringDto.setSaksstatus(fagsak.getStatus());
            fagsakOppsummeringDto.setOpprettetDato(fagsak.getRegistrertDato());
            fagsakOppsummeringDto.setHovedpartRolle(fagsak.getHovedpartRolle());

            List<Behandling> behandlinger = fagsak.hentBehandlingerSortertSynkendePåRegistrertDato();

            List<BehandlingOversiktDto> behandlingOversiktDtoer = behandlinger.stream()
                .map(this::tilBehandlingOversiktDto)
                .toList();

            fagsakOppsummeringDto.setNavn(hentNavn(behandlinger));
            fagsakOppsummeringDto.setBehandlingOversikter(behandlingOversiktDtoer);
            fagsakListe.add(fagsakOppsummeringDto);
        }
        return fagsakListe;
    }

    private BehandlingOversiktDto tilBehandlingOversiktDto(Behandling behandling) {
        BehandlingOversiktDto behandlingOversiktDto = new BehandlingOversiktDto();
        if (behandling != null) {
            Behandlingsresultat behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandling.getId());

            behandlingOversiktDto.setBehandlingID(behandling.getId());
            behandlingOversiktDto.setBehandlingsstatus(behandling.getStatus());
            behandlingOversiktDto.setBehandlingstype(behandling.getType());
            behandlingOversiktDto.setBehandlingstema(behandling.getTema());
            behandlingOversiktDto.setOpprettetDato(behandling.getRegistrertDato());
            behandlingOversiktDto.setBehandlingsresultattype(behandlingsresultat.getType());
            behandlingOversiktDto.setSvarFrist(behandling.getDokumentasjonSvarfristDato());

            setPeriodeOpplysninger(behandling, behandlingOversiktDto);
        }
        return behandlingOversiktDto;
    }

    private void setPeriodeOpplysninger(Behandling behandling, BehandlingOversiktDto behandlingOversiktDto) {
        saksopplysningerService.finnSedOpplysninger(behandling.getId()).ifPresentOrElse(sedDoument -> {
                var land = SoeknadslandDto.av(sedDoument.getLovvalgslandKode());
                behandlingOversiktDto.setLand(land);

                var periode = sedDoument.getLovvalgsperiode();
                behandlingOversiktDto.setSoknadsperiode(new PeriodeDto(periode.getFom(), periode.getTom()));
            },
            () -> {
                var mottatteOpplysninger = mottatteOpplysningerService.finnMottatteOpplysninger(behandling.getId());
                if (mottatteOpplysninger.isPresent()) {
                    var mottatteOpplysningerData = mottatteOpplysninger.get().getMottatteOpplysningerData();

                    var land = SoeknadslandDto.av(hentLand((mottatteOpplysningerData)));
                    behandlingOversiktDto.setLand(land);

                    var periode = hentPeriode(mottatteOpplysningerData);
                    if (periode != null) {
                        behandlingOversiktDto.setSoknadsperiode(new PeriodeDto(periode.getFom(), periode.getTom()));
                    }
                }
            });

        Behandlingsresultat behandlingsResultat = behandlingsresultatService.hentResultatMedMedlemskapOgLovvalg(behandling.getId());

        behandlingsResultat.finnLovvalgsperiode().ifPresent(lovvalgsperiode -> {
            var periode = new PeriodeDto(lovvalgsperiode.getFom(), lovvalgsperiode.getTom());
            behandlingOversiktDto.setLovvalgsperiode(periode);
        });

        var periode = new PeriodeDto(
            behandlingsResultat.utledMedlemskapsperiodeFom(),
            behandlingsResultat.utledMedlemskapsperiodeTom());
        behandlingOversiktDto.setMedlemskapsperiode(periode);
    }

    private String hentNavn(List<Behandling> behandlinger) {
        if (behandlinger.isEmpty()) {
            return UKJENT_NAVN;
        }
        var fagsak = behandlinger.get(0).getFagsak();
        var aktørId = fagsak.finnBrukersAktørID();
        if (aktørId != null) {
            return persondataFasade.hentSammensattNavn(persondataFasade.hentFolkeregisterident(aktørId));
        }
        var orgnr = fagsak.finnVirksomhetsOrgnr();
        if (orgnr != null) {
            return organisasjonOppslagService.hentOrganisasjon(orgnr).getNavn();
        }
        return UKJENT_NAVN;
    }
}

File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/fagsaker/FagsakController.kt
package no.nav.melosys.tjenester.gui.fagsaker

import io.swagger.annotations.Api
import io.swagger.annotations.ApiOperation
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Fagsak
import no.nav.melosys.domain.Lovvalgsperiode
import no.nav.melosys.domain.kodeverk.Aktoersroller
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper
import no.nav.melosys.domain.util.MottatteOpplysningerUtils
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.service.behandling.BehandlingsresultatService
import no.nav.melosys.service.mottatteopplysninger.MottatteOpplysningerService
import no.nav.melosys.service.persondata.PersondataFasade
import no.nav.melosys.service.registeropplysninger.OrganisasjonOppslagService
import no.nav.melosys.service.sak.*
import no.nav.melosys.service.saksopplysninger.SaksopplysningerService
import no.nav.melosys.service.tilgang.Aksesskontroll
import no.nav.melosys.sikkerhet.context.SubjectHandler
import no.nav.melosys.tjenester.gui.dto.BehandlingOversiktDto
import no.nav.melosys.tjenester.gui.dto.FagsakDto
import no.nav.melosys.tjenester.gui.dto.FagsakOppsummeringDto
import no.nav.melosys.tjenester.gui.dto.FagsakSokDto
import no.nav.melosys.tjenester.gui.dto.SoeknadslandDto.Companion.av
import no.nav.melosys.tjenester.gui.dto.periode.PeriodeDto
import no.nav.security.token.support.core.api.Protected
import org.apache.commons.lang3.StringUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Scope
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.context.WebApplicationContext
import java.util.function.Consumer

@Protected
@RestController
@RequestMapping("/fagsaker")
@Api(tags = ["fagsaker"])
@Scope(value = WebApplicationContext.SCOPE_REQUEST)
class FagsakController(
    private val fagsakService: FagsakService,
    private val aksesskontroll: Aksesskontroll,
    private val mottatteOpplysningerService: MottatteOpplysningerService,
    private val opprettSak: OpprettSak,
    private val endreSakService: EndreSakService,
    private val behandlingsresultatService: BehandlingsresultatService,
    private val persondataFasade: PersondataFasade,
    private val saksopplysningerService: SaksopplysningerService,
    private val organisasjonOppslagService: OrganisasjonOppslagService,
    private val opprettBehandlingForSak: OpprettBehandlingForSak,
    private val ferdigbehandleService: FerdigbehandleService
) {
    @GetMapping("/{saksnr}")
    @ApiOperation(
        value = "Henter en sak med et gitt saksnummer",
        notes = ("Spesifikke saker kan hentes via saksnummer.")
    )
    fun hentFagsak(@PathVariable("saksnr") saksnummer: String?): ResponseEntity<FagsakDto?> {
        val fagsak = fagsakService.hentFagsak(saksnummer)
        aksesskontroll.autoriserSakstilgang(fagsak)
        val fagsakDto = tilFagsakDto(fagsak)
        log.info("Henting av sak {} ({})", fagsakDto.saksnummer, fagsakDto.gsakSaksnummer)
        return ResponseEntity.ok(fagsakDto)
    }

    @PostMapping
    @ApiOperation(value = "Oppretter en ny sak.")
    fun opprettNySak(@RequestBody opprettSakDto: OpprettSakDto): ResponseEntity<Void?> {
        if (opprettSakDto.brukerID == null && opprettSakDto.virksomhetOrgnr == null) {
            throw FunksjonellException("BrukerID eller organisasjonsnummer trengs for å opprette en sak.")
        }
        if (opprettSakDto.brukerID != null) {
            aksesskontroll.autoriserFolkeregisterIdent(opprettSakDto.brukerID)
        }

        if (opprettSakDto.oppgaveID == null) {
            opprettSak.opprettNySakOgBehandling(opprettSakDto)
        } else {
            opprettSak.opprettNySakOgBehandlingFraOppgave(opprettSakDto)
        }

        return ResponseEntity.noContent().build<Void?>()
    }

    @PostMapping("/{saksnr}/behandlinger")
    @ApiOperation(value = "Oppretter en ny behandling for sak.")
    fun opprettNyBehandlingForSak(
        @PathVariable("saksnr") saksnummer: String?,
        @RequestBody opprettSakDto: OpprettSakDto
    ): ResponseEntity<Void?> {
        if (opprettSakDto.brukerID != null) {
            aksesskontroll.autoriserFolkeregisterIdent(opprettSakDto.brukerID)
        }

        opprettBehandlingForSak.opprettBehandling(saksnummer, opprettSakDto)

        return ResponseEntity.noContent().build<Void?>()
    }

    @PutMapping("/{saksnr}")
    @ApiOperation(value = "Endre en sak.")
    fun endreFagsak(
        @PathVariable("saksnr") saksnummer: String,
        @RequestBody endreDto: EndreSakDto
    ): ResponseEntity<Void?> {
        log.debug(
            "Saksbehandler {} ber om å endre fagsak {} med sakstype {}, sakstema {}",
            SubjectHandler.getInstance().getUserID(), saksnummer, endreDto.sakstype, endreDto.sakstema
        )
        aksesskontroll.autoriserSakstilgang(saksnummer)

        if (endreDto.behandlingstype == Behandlingstyper.ÅRSAVREGNING && endreDto.behandlingID != null) {
            endreSakService.endreÅrsavregningBehandling(
                endreDto.behandlingID!!,
                endreDto.behandlingsstatus,
                endreDto.mottaksdato
            )
        } else {
            endreSakService.endre(
                saksnummer, endreDto.sakstype, endreDto.sakstema, endreDto.behandlingstema,
                endreDto.behandlingstype, endreDto.behandlingsstatus, endreDto.mottaksdato
            )
        }

        return ResponseEntity.noContent().build<Void?>()
    }

    @PostMapping("/sok")
    @ApiOperation(
        value = "Søk etter saker på ident eller saksnummer eller orgnr",
        notes = ("Saker knyttet til en bruker søkes via fødselsnummer eller d-nummer. Saker knyttet til en organisasjon søkes via organisasjonsnummer."),
        response = FagsakOppsummeringDto::class,
        responseContainer = "List"
    )
    fun hentFagsaker(@RequestBody fagsakSokDto: FagsakSokDto): MutableList<FagsakOppsummeringDto?> {
        if (StringUtils.isNotEmpty(fagsakSokDto.ident)) {
            aksesskontroll.auditAutoriserFolkeregisterIdent(
                fagsakSokDto.ident,
                "Søk på person med ident. Oversikt over saker og behandlinger."
            )
            return tilFagsakOppsummeringDtoer(
                fagsakService.hentFagsakerMedAktør(
                    Aktoersroller.BRUKER,
                    fagsakSokDto.ident
                )
            )
        } else if (StringUtils.isNotEmpty(fagsakSokDto.saksnummer)) {
            val fagsak = fagsakService.finnFagsakFraSaksnummer(fagsakSokDto.saksnummer)
            if (fagsak.isPresent) {
                aksesskontroll.auditAutoriserSakstilgang(
                    fagsak.get(),
                    "Søk på sak med saksnummer. Oversikt over saker og behandlinger."
                )
                return tilFagsakOppsummeringDtoer(listOf<Fagsak>(fagsak.get()))
            }
        } else if (StringUtils.isNotEmpty(fagsakSokDto.orgnr)) {
            return tilFagsakOppsummeringDtoer(
                fagsakService.hentFagsakerMedOrgnr(
                    Aktoersroller.VIRKSOMHET,
                    fagsakSokDto.orgnr
                )
            )
        }

        return mutableListOf<FagsakOppsummeringDto?>()
    }

    @PutMapping("/{behandlingID}/ferdigbehandle")
    @ApiOperation("Avslutt behandling med Ferdigbehandlet som resultat og oppdatere saksstatus")
    fun ferdigbehandleSak(@PathVariable("behandlingID") behandlingID: Long): ResponseEntity<Void?> {
        log.info(
            "Saksbehandler {} ber om å ferdigbehandle behandling {}",
            SubjectHandler.getInstance().getUserID(),
            behandlingID
        )
        aksesskontroll.autoriserSkriv(behandlingID)

        ferdigbehandleService.ferdigbehandle(behandlingID)

        return ResponseEntity.noContent().build<Void?>()
    }


    private fun tilFagsakDto(fagsak: Fagsak): FagsakDto {
        val fagsakDto = FagsakDto()
        fagsakDto.saksnummer = fagsak.saksnummer
        fagsakDto.gsakSaksnummer = fagsak.gsakSaksnummer
        fagsakDto.sakstema = fagsak.tema
        fagsakDto.sakstype = fagsak.type
        fagsakDto.saksstatus = fagsak.status
        fagsakDto.registrertDato = fagsak.getRegistrertDato()
        fagsakDto.endretDato = fagsak.endretDato
        fagsakDto.hovedpartRolle = fagsak.hovedpartRolle

        return fagsakDto
    }

    private fun tilFagsakOppsummeringDtoer(saker: List<Fagsak>): MutableList<FagsakOppsummeringDto?> {
        val fagsakListe: MutableList<FagsakOppsummeringDto?> = ArrayList<FagsakOppsummeringDto?>()
        for (fagsak in saker) {
            val fagsakOppsummeringDto = FagsakOppsummeringDto()
            fagsakOppsummeringDto.saksnummer = fagsak.saksnummer
            fagsakOppsummeringDto.sakstema = fagsak.tema
            fagsakOppsummeringDto.sakstype = fagsak.type
            fagsakOppsummeringDto.saksstatus = fagsak.status
            fagsakOppsummeringDto.opprettetDato = fagsak.getRegistrertDato()
            fagsakOppsummeringDto.hovedpartRolle = fagsak.hovedpartRolle

            val behandlinger = fagsak.hentBehandlingerSortertSynkendePåRegistrertDato()

            val behandlingOversiktDtoer = behandlinger.map { behandling -> tilBehandlingOversiktDto(behandling) }

            fagsakOppsummeringDto.navn = hentNavn(behandlinger)
            fagsakOppsummeringDto.behandlingOversikter = behandlingOversiktDtoer
            fagsakListe.add(fagsakOppsummeringDto)
        }
        return fagsakListe
    }

    private fun tilBehandlingOversiktDto(behandling: Behandling?): BehandlingOversiktDto {
        val behandlingOversiktDto = BehandlingOversiktDto()
        if (behandling != null) {
            val behandlingsresultat = behandlingsresultatService.hentBehandlingsresultat(behandling.id)

            behandlingOversiktDto.behandlingID = behandling.id
            behandlingOversiktDto.behandlingsstatus = behandling.status
            behandlingOversiktDto.behandlingstype = behandling.type
            behandlingOversiktDto.behandlingstema = behandling.tema
            behandlingOversiktDto.opprettetDato = behandling.getRegistrertDato()
            behandlingOversiktDto.behandlingsresultattype = behandlingsresultat.type
            behandlingOversiktDto.svarFrist = behandling.dokumentasjonSvarfristDato

            setPeriodeOpplysninger(behandling, behandlingOversiktDto)
        }
        return behandlingOversiktDto
    }

    private fun setPeriodeOpplysninger(behandling: Behandling, behandlingOversiktDto: BehandlingOversiktDto) {
        saksopplysningerService.finnSedOpplysninger(behandling.id).let { sedOptional ->
            if (sedOptional.isPresent) {
                sedOptional.get().let { sed ->
                    behandlingOversiktDto.apply {
                        land = av(sed.lovvalgslandKode)
                        soknadsperiode = PeriodeDto(sed.lovvalgsperiode.fom, sed.lovvalgsperiode.tom)
                    }
                }
            } else {
                mottatteOpplysningerService.finnMottatteOpplysninger(behandling.id).ifPresent { mottatteOpplysninger ->
                    mottatteOpplysninger.mottatteOpplysningerData.let { data ->
                        behandlingOversiktDto.apply {
                            land = av(MottatteOpplysningerUtils.hentLand(data))
                            MottatteOpplysningerUtils.hentPeriode(data)?.let { periode ->
                                soknadsperiode = PeriodeDto(periode.fom, periode.tom)
                            }
                        }
                    }
                }
            }
        }


        val behandlingsResultat = behandlingsresultatService.hentResultatMedMedlemskapOgLovvalg(behandling.id)

        behandlingsResultat.finnLovvalgsperiode().ifPresent(Consumer { lovvalgsperiode: Lovvalgsperiode? ->
            val periode = PeriodeDto(lovvalgsperiode!!.fom, lovvalgsperiode.tom)
            behandlingOversiktDto.lovvalgsperiode = periode
        })

        val periode = PeriodeDto(
            behandlingsResultat.utledMedlemskapsperiodeFom(),
            behandlingsResultat.utledMedlemskapsperiodeTom()
        )
        behandlingOversiktDto.medlemskapsperiode = periode
    }

    private fun hentNavn(behandlinger: List<Behandling>): String? {
        if (behandlinger.isEmpty()) {
            return UKJENT_NAVN
        }
        val fagsak = behandlinger.get(0)!!.fagsak
        val aktørId = fagsak.finnBrukersAktørID()
        if (aktørId != null) {
            return persondataFasade.hentSammensattNavn(persondataFasade.hentFolkeregisterident(aktørId))
        }
        val orgnr = fagsak.finnVirksomhetsOrgnr()
        if (orgnr != null) {
            return organisasjonOppslagService.hentOrganisasjon(orgnr).navn
        }
        return UKJENT_NAVN
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(FagsakController::class.java)
        private const val UKJENT_NAVN = "UKJENT"
    }
}
